### PR TITLE
Makefile - Detect MKLROOT for LIBXSMM linking

### LINKED ISSUE #199: Linking BLAS
On CU Summit using the Intel compilers, the libXSMM backend requires linking against MKL instead of BLAS. Currently this requires a manual Makefile edit.

### PR DESCRIPTION
**Pull request by jeremylt**
Fixes #199

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -292,10 +292,12 @@ ifneq ($(wildcard $(XSMM_DIR)/lib/libxsmm.*),)
...
   MKL ?= 0
-  ifneq (0,$(MKL))
-    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
-  else
+  ifeq (0,$(MKL))
+  ifeq ($(MKLROOT),)
     BLAS_LIB = -lblas
+  else
+    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
+  endif
```
Comment (jedbrown) [inline Makefile]: Else...

File: Makefile [hunk 2]
```diff
@@ -292,10 +292,12 @@ ifneq ($(wildcard $(XSMM_DIR)/lib/libxsmm.*),)
...
   $(libceeds) : LDLIBS += -lxsmm -ldl
   MKL ?= 0
-  ifneq (0,$(MKL))
-    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
-  else
+  ifeq (0,$(MKL))
+  ifeq ($(MKLROOT),)
     BLAS_LIB = -lblas
+  else
+    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```
Comment (jedbrown) [inline Makefile]: Add -L and rpath flags.

File: Makefile (hunk 1 above)
Comment (jeremylt) [reply Makefile]: Found a better way to do it.

File: Makefile (hunk 2 above)
Comment (jeremylt) [reply Makefile]: For MKL? We haven't needed those flags on CU's machine in the past.

**Comment by jedbrown** [comment]
Probably because you sourced something that put it in the LIBRARY_FLAGS and LD_LIBRARY_PATH environment variables.On Sep 17, 2019 18:22, Jeremy L Thompson <notifications@github.com> wrote:@jeremylt commented on this pull request.



In Makefile:
>      BLAS_LIB = -lblas
+  else
+    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl

For MKL? We haven't needed those flags on CU's machine in the past.

**Comment by jeremylt** [comment]
Hmm, Intel's magic incantation website only lists -L
``` 
-L${MKLROOT}/lib/intel64 -Wl,--no-as-needed -lmkl_intel_ilp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```
So you're suggesting we also add 
`-rpath,$(abspath $(MKLROOT)/lib/intel64`?

**Comment by jeremylt** [comment]
Added as two separate commits so we can drop the -rpath flag if desired.

**Comment by jeremylt** [comment]
I verified that this branch works as is on CU's machine.

File: Makefile [hunk 3]
```diff
@@ -291,11 +291,11 @@ endif
...
   $(libceeds) : LDLIBS += -lxsmm -ldl
-  MKL ?= 0
-  ifneq (0,$(MKL))
-    BLAS_LIB = -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
-  else
+  MKL ?=
+  ifeq (,$(MKL)$(MKLROOT))
     BLAS_LIB = -lblas
+  else
+    BLAS_LIB = -L$(MKLROOT)/lib/intel64 -Wl,-rpath,$(abspath $(MKLROOT)/lib/intel64),--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```
Comment (jedbrown) [inline Makefile]: If MKLROOT was empty, then you could be passing `-L/lib/intel64`, which may not exist, and thus produce errors. Perhaps something like this?
```suggestion
    BLAS_LIB = $(if $(MKLROOT),-L$(MKLROOT)/lib/intel64 -Wl,-rpath,$(MKLROOT)/lib/intel64) -Wl,--no-as-needed -lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl
```

Comment (jeremylt) [reply Makefile]: All set

**Pull request merged**